Paper: Supporting Greater Interactivity in the IPython Visualization Ecosystem (#912)
scipy-conference/scipy_proceedings: papers/nathan_martindale/main.md

visualization ecosystem within Python. Although a great deal of progress has
been made in this community, a missing element is a more generalized and
flexible approach to aspects of the [details-on-demand](#details-on-demand)
paradigm. To support this approach, the authors have developed a library that
provides wrapper components that can render any widget as an overlay on top of
other widgets. These components simplify designing layouts and visualizations
that would be difficult to implement from scratch.

In this work, we first discuss IPyWidgets and what makes it such an important
piece of the overall system, as well as some of the limitations encountered when
building only with the base IPyWidgets framework. We then briefly cover some of
the available libraries that build on top of IPyWidgets, highlighting
IPyVuetify, which addresses most of the limitations of base IPyWidgets. Finally,
we discuss a library we have open sourced called IPyOverlay and an example
use case that motivated its design. 

### Details on Demand

Details on demand is one piece of the "visual information-seeking mantra"
[@shneidermanEyesHaveIt1996], a paper describing an overall design approach to

> building interactive tooling. The general flow of interaction proposed by this

Reviewer: This mention of details on demand should really refer to the AUTHOR and not "a paper".
```suggestion
Details on demand is one piece of the "visual information-seeking mantra" of Schneiderman (1996)
 describing an overall design approach to building interactive tooling. 
 The general flow of interaction proposed by this
```

mantra is "overview first, zoom and filter, then details-on-demand"
[@shneidermanEyesHaveIt1996]. Overview

> first ideally starts with broader aggregate visualizations: something to gain a sense

Reviewer: Redundant `first`
```suggestion
ideally starts with broader aggregate visualizations: something to gain a sense
```

of the general scope, complexity, or shape of the data. From there a user can

> identify their own areas of interest to hone in on with whatever zoom and filter

Reviewer: Redundant since user unlikely to identify other's areas of interest:
```suggestion
identify areas of interest to hone in on with whatever zoom and filter
```

tools are applicable for their data's modality. Finally, details on demand
means that the user can see or interact with more in-depth and specific
information for pieces of their data, often in the form of a pop-up window or

> similar. This selective detail capability provides a flexible user experience

Reviewer: I do not like the phrasing "pop-up window or similar" since it doesn't say "similar WHAT".  Maybe it would be better to say "in the form of something like a pop-up window or inset graph"?  I just find the current phrasing gets in the way to understanding the authors' meaning.

that keeps the initial interface uncluttered without sacrificing the ability to
organically explore the data.

A simple and widespread example of details on demand are tooltips. Descriptions
of components or data point values in visualizations would often create an
unusably noisy interface if they were always visible. Instead, a more concise
view is maintained while providing the necessary information in a spatially
relevant view when the user specifically requests it. Readers viewing the web

> version of these proceedings can explore a simple example of details on demand

Reviewer 2: Do we see this "web version" in the curvenote preview?  Is this going to be a separate file?

Author: Yeah, the curvenote preview is the web version, so hovering over the "Figure 1" link pulls up a tooltip/details-on-demand example. That figure itself is then just a (almost recursive :grin:) screenshot of what hovering over the figure link looks like so that you still get the idea if reading on a non-web medium, like a PDF version of the paper.

Reviewer 2: that makes sense, cool

live by hovering over any of the figure references.
@fig:detailsondemand shows how hovering over a reference (the demand)
opens a window at the cursor that displays the target of the reference (the
details).

:::{figure} detailsondemand.png 
:label: fig:detailsondemand 
A figure demonstrating details on demand in action when a user hovers over the figure
reference for @fig:detailsondemand. 
:::

## IPyWidgets

IPyWidgets is a library that enables implementing user interactivity inside of a

> Jupyter Notebook/Jupyter Lab. A key challenge this framework has to overcome is

Reviewer: I don't think you need to specify both Jupyter Notebook and Jupyter Lab since Jupyter notebooks (the documents) run in different implementations of Jupyter (including Jupyter in VS Code).  I would suggest just
```suggestion
Jupyter Notebook. A key challenge this framework has to overcome is
```

import ipywidgets as ipw

# initialize the widgets
graph1_out = ipw.Output()
graph2_out = ipw.Output()

parameter1 = ipw.FloatSlider(...)
parameter2 = ipw.IntText(...)
parameter3 = ipw.IntSlider(...)

# create the dashboard layout, combining all the widgets
dashboard = ipw.VBox([
    ipw.HBox([graph1_out, graph2_out]),
    ipw.HBox([parameter1, parameter2, parameter3]),
])

# event handler functions
def on_parameter1_change(change):
    with graph1_out:
        graph1_render_function(parameter1.value, parameter2.value)
        
def on_parameter2_change(change):
    with graph1_out:
        graph1_render_function(parameter1.value, parameter2.value)
    with graph2_out:
        graph2_render_function(parameter2.value, parameter3.value)
        
def on_parameter3_change(change):
    with graph2_out:
        graph2_render_function(parameter2.value, parameter3.value)

# attach event handlers
parameter1.observe(on_parameter1_change, names=["value"])
parameter2.observe(on_parameter2_change, names=["value"])
parameter3.observe(on_parameter3_change, names=["value"])

> ```

Reviewer 2: This example is pretty generic.  Could you change the names to make it have more meaningful function.  This would be easier to understand.

Author: Yeah that's a good point, I renamed things to try to make it more obvious that this is showing a pattern where you can have parameters that influence multiple graphs, or ipywidget outputs that are influenced by multiple inputs that can be shared or have crossover

### State Synchronization

One of the primary challenges that IPyWidgets helps solve is the problem of
state synchronization, where updates to variables made on either side of the
Python-JavaScript link are monitored and communicated across when they occur.
Based on this communication, changes to variables in Python are reflected in the
browser's JavaScript and vice versa. Under the hood, IPyWidgets makes use of
backbone.js [@LowLevelWidget], a model-view-* (MV*) framework for JavaScript
that implements synchronization between a JavaScript _model_ (where
data/attributes/values for a widget are stored) and a JavaScript _view_ (the
code that creates the visual UI elements the user sees and interacts with in
their browser). IPyWidgets connects to this backbone model via WebSockets,
through which model changes and events are sent as serialized JSON.

:::{figure} ipywidgets_state.png
:label: fig:ipywidgets_state
State mirroring between models and view interaction for an example slider widget.

> :::

Reviewer 2: The blue color in the bottom right arrow seems off.  Maybe you could make the bottom box a different shade of yellow and change the arrow as well.

Author: As in the blue arrow color clashes with the shade of yellow we're using or as in that arrow shouldn't be blue?

Our intent was to show the two pathways of changes colored based on source, so:
1. Interacting with the HTML/JS causes changes to propagate from the JS view -> JS model -> python model (represented as yellow arrows the whole way to match the JS yellow) 
2. Modifying values from python propagates to the view in the other direction, so the pathway is python model -> JS model -> JS view (which we represented with blue arrows the whole way to match the python blue)

I'm assuming the oddness is having the python blue arrow between two yellow JS boxes, and the idea was to show that that communication is being caused by a change on the python side. Does that make more sense or is there maybe a better way to represent this?

Reviewer 2: gotcha, yeah, I guess a third color might be too confusing

Reviewer: I'll note that for accessibility purposes (e.g. - usability by the color blind), the color should not be the only thing distinguishing the arrows if you want them to be distinct.  I don't have the figure here, but I thought I would mention it.  Doesn't directly address the question at hand, but just something I forgot to mention when originally reviewing the figure.

@fig:ipywidgets_state shows a simplified view of the communication channels
between the Python model and the JavaScript model/view for an example slider widget.
Modifying an attribute on the Python model sends the modification to the
JavaScript model, where any changes that would modify the visual aspects of the
widget are communicated to the JavaScript view (e.g., a change in the slider
value would correspondingly update the position of the slider indicator within
the track). Similarly, a user interacting with the widget (e.g., moving the
slider within the track) updates the JavaScript state, which communicates this
change to the Python model and calls any event handlers listening for
changes to the value.

### Common Framework

One of the most important benefits of IPyWidgets is that it provides a low-level
framework abstracting away many of the specifics for how communication between a
kernel, server, and frontend takes place. This framework can be used to design
new widgets by providing implementations for the models and view code, without
needing to reinvent the various messaging system mechanisms that solve the state

> synchronization problem discussed above. All of this has allowed many libraries

Reviewer: This is a pet peeve of mine, I don't think it is *necessary* to fix, but I do not like the use of spatial terms like "above" for prose, I prefer time terms like "previously".  But I do believe this spatial language is commonly accepted now, I suppose I should not fight it.
```suggestion
synchronization problem discussed previously. All of this has allowed many libraries
```

Author: Yeah that's totally fair, especially for like a PDF print I suppose "above" kinda loses any meaning if "above" is now an entirely different page! Fixed.

nested to allow grouping widgets. 
Although these components are often more than sufficient for simple visualizations,
more complex interfaces with richer interaction paradigms can be hard to achieve.
Some examples of dashboard elements that cannot be created
with IPyWidgets' default components include interactive data frames, dialog
windows, and floating buttons.

#### Events

Another limiting factor for creating more complex interactivity is the type of
events that can be observed on the components. UIs are inherently
event-driven; the user interacts with widgets/visualizations, and these events
run code that update the interface. This is most often
implemented by registering "observers" or event-handlers, which tells IPyWidgets to
run a specified function every time a particular event is raised. In practice,
for the default components provided by the IPyWidgets library, this is mostly limited to
value changes (e.g., a slider is moved, the text in a textbox is changed, or an
item in a drop-down menu is selected).

These types of events are sufficient for basic interfaces, but often more complex types of

> interactions or responsiveness can greatly enhance the capabilities of a tool.

Reviewer: You kind of wedge in "or responsiveness" in this sentence and it made it hard to follow.  Furthermore, you address this later in the text whereas this paragraph focuses exclusively on other interactions like mouse events, just keep it on that:
```suggestion
interactions can greatly enhance the capabilities of a tool.
```

subset of rules available through Python attributes. This makes common
style changes easier to work with, but the lack of any direct access to arbitrary CSS
rules limits more complex layouts, namely the ability to float
one widget on top of another.

#### Custom IPyWidget Complexity

A key benefit of IPyWidgets as a framework is that it allows
developers to design, build, and publish new components that other people can
use in place of or alongside the default widgets. However, the complexity of the
infrastructure required to do this with the base IPyWidgets library still makes
it challenging to prototype or put together a one-off custom component in a
project. Several libraries we discuss in the next section have improved this
process substantially, but traditionally a custom component library was
developed by starting from a Jupyter-provided cookie-cutter template
[@ipywidgets_cookiecutter] in a separate project.

One of the difficulties that comes with a project created in this way is the
need to include the development infrastructure of both a Python and a JavaScript
project. JavaScript toolchains and dependency management are somewhat

> notoriously complicated. Prototyping the component can be a slow process as

Reviewer: Can you provide a citation for "somewhat notoriously complicated"?  If it is "notorious" it suggests it is "well known" and thus should be citable.  If not, consider rewording this.  I think the existence of cookie-cutter templates definitely suggests this is a complex task.

Author: Alas, my citations would probably only consist of reddit programming memes :grinning: I do think this is a fairly commonly pointed out issue, but I don't really have a good formal reference for it, so point taken! I reworded it instead to briefly explain why it can be complicated:
```
JavaScript toolchains and dependency management can quickly become 
complicated with the number of packages, build system tools, and config files 
often necessary.
```
(let me know if you think this is sufficient)

relative ease. From general plotting tools like interactive Matplotlib
[@10.5281/zenodo.10974323]), plotly [@plotly], mpld3 [@mpld3], altair
[@altair], and bqplot [@bqplot], to more complex or specialized visualization
widgets like pythreejs [@pythreejs] and ipyvolume [@10.5281/zenodo.1286976] for
3D rendering, ipyleaflet [@ipyleaflet] for interactive maps, and ipycytoscape
[@ipycytoscape] for interactive networks, there are dozens of packages providing
specific components and bridges to JavaScript visualization libraries. Other
libraries such as Panel, Solara, and Streamlit provide more holistic
integrations between IPyWidgets and other visualization techniques, 
which can support production-style dashboards for
use outside of Jupyter alone. Although these libraries address the limited component
set included with IPyWidgets, another class of Python packages exists in this ecosystem
that completely changes and improves the workflow for defining new widgets. These
workflow improvements eliminate the need for JavaScript toolchains, and in some cases, they also
provide the capabilities from various JavaScript web frameworks. A few

> we briefly touch on here are AnyWidget, IPyReact, and IPyVuetify.

Reviewer 2: Gradio probably deserves a mention too.

Author: Gradio is a good suggestion, I looked into it and was kind of surprised to see it doesn't actually have any direct ipywidgets compatibility. It has its own set of communication mechanisms, and while it can render inside of a notebook, as far as I can tell it's still entirely distinct and can't use other ipywidgets

AnyWidget is used in another project called IPyReact [@ipyreact], which brings
the ReactJS library to support component-based design and allows wrapping
many existing React components.

A similar library we focus on in this work that addresses all of the [main
limitations of IPyWidgets](#limitations) is IPyVuetify [@ipyvuetify], which
brings many of the capabilities of Vue.js into IPyWidgets. Vue.js is a web
component framework similar to React and emphasises a component-driven design,
where a single Vue file can contain an entire component: the template HTML to
render, the CSS to apply, and the JavaScript to provide any interactivity
and mechanisms to connect it to the rest of the application. Attributes within
the HTML can be bound to JavaScript variables, which makes it possible to easily
automate the JavaScript model $\rightarrow$ view updates. IPyVuetify brings two main
advantages: a large library of flexible premade components and the ability to
easily define new components on the fly using Vue to simplify the resulting
JavaScript code.

### Vuetify Component Library

IPyVuetify is an IPyWidgets port of the Vuetify component library, a set of more

> than 70 premade material design UI widgets. Most of these components support a

Reviewer: Can you define "material design" or cite the term?

Author: Provided a reference to google's material design website

fairly extensive set of attributes and properties to change both appearance and
functionality, and more importantly each generally has multiple types of events

> that can be observed. A simple example would be a textbox form input component.

Reviewer: Run on sentence with two ideas:  the support of extensive attributes/properties and multiple types of events to be observed.  Split it up:
```suggestion
functionality.  More importantly, each iPyVuetify widget generally has multiple types of events
that can be observed. A simple example would be a textbox form input component.
```

In IPyWidgets, the `Text` widget can mostly only observe a value change when the
user changes the text inside. The IPyVuetify `TextField` can additionally raise
events for key presses, mouse clicks, the element gaining and losing focus, as
well as any HTML event. This level of control can be useful, for example, when a
text value change triggers a heavy processing task. The interaction is more

> effective when the computation is delayed until the field loses focus or the

Reviewer: Here I think the proper word is "responsive" instead of "effective":
```suggestion
responsive when the computation is delayed until the field loses focus or the
```

user has pressed enter, rather than restarting the computation on every letter
the user adds or removes.

All of the default components in IPyWidgets similarly have more flexible
versions and many additional components in IPyVuetify. @fig:ipw_ipv 
shows a small sample of default IPyWidgets on the left with their corresponding
IPyVuetify versions on the right. Also pictured at the bottom is the Vuetify
DataTable, which is an example of a component that is not in the default IPyWidgets. The DataTable allows 
interactivity to be added to a table supporting filtering, searching, and both client/server 
side pagination. Other examples of useful UI elements IPyVuetify adds 
are various layout widgets like navigation drawers, dialog windows, and
steppers.

<!-- would potentially be nice to see a searchbox for table --> 
:::{figure} ipw_ipv.png
:label: fig:ipw_ipv
A comparison of a few IPyWidget components on the upper left and their corresponding
IPyVuetify components on the upper right. The table on the bottom is a
DataTable component, which does not have a default IPyWidgets equivalent.

> :::

Reviewer 2: As in "program 2", this would be more instructive if the buttons meant something.

Author: The figure was just supposed to show the visual distinction between the types of components available in both, so I just labeled them according to the library they came from. I updated the caption to highlight these are just widget samples

the element.

### Custom Vue Components

IPyVuetify provides clean abstractions and mechanisms for creating custom
components. Similar to AnyWidget, IPyVuetify allows defining a Python model
class with synced `traitlets`, all of which are automatically available within
the JavaScript code. In IPyVuetify's case, the frontend code is expected to be a
string of Vue code or a path to a Vue file, which allows for fairly
self-contained HTML, JavaScript, and styling. The design of Vue.js lends itself
well to reducing boilerplate by allowing bindings in the HTML to directly refer
to Python attribute traitlets, removing both the need to specify the attribute
in the JavaScript model as well as the function to update the view from the
model. We discuss this further in the section on [implementing IPyOverlay in
IPyWidgets](#implementing-in-ipywidgets). @code:ipyvuetify exemplifies this,
defining a custom button with a color change on toggle, where the colors can be
modified from Python as well, shown in @fig:ipyvuetify_example. In this
particular example, both the `@click` event and `:color` property directly bind
to the Python attributes, and no explicit JavaScript is necessary. Finally,

> IPyVuetify components make it easy to directly call functions across the

Reviewer: Can you tighten this a bit by stating 
```suggestion
IPyVuetify makes it easy to directly call functions across the
```

In this use case, the dashboard is constructed with IPyWidgets and IPyOverlay,
taking advantage of the techniques described in this section, to determine the
success of training a model to meet our domain-specific criteria. More
specifically, we use Barlow Twins [@barlowtwins] as a self-supervised learning objective
to try to force a model’s learned representations to reflect source properties
rather than other observational characteristics. For example, different
geographic locations will observe slightly different signals of the same source
event based on geological differences in the paths to those locations from the
source. An ideal model would exhibit path invariance and represent these signals similarly. We use the
constructed visualization tool along with metadata (information other than the
timeseries the model was trained on, e.g. event location, depth, distance) to
understand how well the training objective succeeds in learning invariance to
the metadata attributes that are unrelated to the event source properties.
Additionally, we use the dashboard to help end users understand
how well the model generalizes to new sources and to explore uncertainty.

> The dashboard includes a UMAP plot of signals embedded by the model under

Reviewer: Please define what I assume is an acronym, UMAP, the first time you use it.

exploration. Each datapoint is a single observed seismic signal, and these

> data points are then clustered with DBSCAN. <!-- the key benefit from IPyOverlay

Reviewer: Please define DBSCAN the first time you use it.  I know the term is common enough in machine learning but the submissions to the SciPy proceedings are meant to be cross-disciplinary.  You may also want to link to the original 1996 by Ester, Kriegel, Sander, and Xu on the algorithm.

is in allowing .... -->
Metadata attributes aggregated 
by cluster can be explored by clicking within the UMAP ipympl plot, which aids
in determining how much correlation exists for those attributes within that
cluster. Invariant properties would tend to show as spread out distributions,
while attributes that heavily inform a models' embedding would tend to cause
clustering and produce narrow distributions.

### Implementing in IPyWidgets

Within base IPyWidgets, there is no straightforward way of constructing a

> layout in which one widget is overlaid on top of another. Certain other libraries
> have limited functionality for this within their own widget instances, for
> example Matplotlib can display an axes within another axes, as can plotly, and
> many libraries support basic tooltips with arbitrary textual content. Others

Reviewer: Again, several ideas in one run-on sentence that should be separated:
```suggestion
layout in which one widget is overlaid on top of another. Certain other libraries
have limited functionality for this within their own widget instances.  For
example Matplotlib and poorly can display an axes within another axes.  Furthermore,
many libraries support basic tooltips with arbitrary textual content. Others
```

such as IPyVuetify support certain types of limited "dialog" overlays, which can
contain inner widgets and display overtop of the rest of the webpage, but do not
support the same type of arbitrary positioning necessary for some
details-on-demand approaches.

This limitation seems to arise primarily from the conventional layout mechanisms
employed by many widget libraries, which consists of nesting rows and columns
(`HBox` and `VBox` from IPyWidgets) of CSS-flexbox style layouts, or even
stricter grid layouts. <!-- a figure showing the difference we're describing
here -->
These approaches are generally easier to design and work
with, and abstract away or entirely avoid some of the messier CSS layout rules
required to render HTML elements that do not follow an ordered top-to-bottom/left-to-right flow.
Fundamentally, to enable this behavior, we implement
custom IPyWidget containers that use specific CSS rules to control for
nonlinear layouts, namely relying on absolute positioning (by setting
`position: absolute`), and each child widget setting `left` and `top` CSS
properties to have pixel-level control of its location relative to the parent

> container. A diagram demonstrating this design principle is shown in

Reviewer: I was confused by this sentence, specifically what is "this behavior"... Do you mean the ability to use the easier `HBox` and `VBox` grid layout or from the context of the sentence, the ability to have absolute positioning and (while you don't mention it) layers!  I would rework this to make it clear from the beginning what behavior you are referring to.

A key enabler is IPyVuetify's ability to embed other IPyWidgets inside
of the template HTML, allowing us to construct custom layouts of one or more
child widgets, similar to `HBox` and `VBox` in IPyWidgets. In effect, we can
wrap any other widget with custom HTML, CSS, and JavaScript. The Vue framework
is particularly well suited to concise implementation with this approach.
Vue-flavored template HTML supports binding reactive properties on attributes
(e.g., `<div :width="some_JS_variable_name">`), meaning we can reference any
model variables in the template syntax, and any time those attributes update,
the corresponding HTML is automatically updated without other code having to monitor and
orchestrate the JavaScript model $\rightarrow$ view update. Additionally, Vue's
template syntax supports several programmatic control structures such as looping
(`v-for`) and conditionals (`v-if`), all of which combined allows
declaratively constructing a bound list of child widgets each wrapped in custom
HTML that reactively responds to list updates and other model property changes.
A simplified example is shown in @code:container.

```{code-block} html
:label: code:container
:linenos: true

> :caption: A simplified version of the Vue template for `OverlayContainer`. The `<div> tag with `v-for` reactively produces a `<div>` tag for each widget in the `children` list attribute within the Python model. Note, the style attribute is bound and references the position arrays on lines 21--22. Changing the position of each overlay widget is done by modifying the corresponding values in these arrays. We keep these arrays only on the JavaScript side to prevent undue traffic between the Python and JavaScript models from high-frequency events (e.g., mouse movements) that can cause significant lag in the interface.

Reviewer: This caption gets mangled when rendered from MyST, I think you forgot to close the quote around the first `<div>` and that mucked it up, but check.

Author: I sure did, that fixed it, thanks! (@ritwit I assume this is what your comment was about as well?)

to specific pieces of the background widget (@fig:overlays_connections).

:::{figure} overlays_connections.png
:label: fig:overlays_connections
The same detail plots as in @fig:overlays. The overlay widgets have been dragged
further out to show the connection lines to the clusters in the UMAP each is based
on.
:::

One end point of the line is always associated with a specified overlay widget
and is automatically updated to the widget's center whenever it is moved. The
other side of the line can be "attached" in a few different ways, including
simply specifying pixel coordinates within the background, but a significant
challenge is locations within dynamic content such as interactive plots that can
be panned and scaled. The conceptual approach to solving this for many plotting
libraries is to store the x/y data point the connection should point to, the
initial ranges of any plot axes, and current pixel boundaries of the plot within
the container. Then, listening for any events that indicate an axis range has
changed, handler functions can recalculate pixel locations based on the new
relative position of the x/y data point within the new range.

> Unfortunately, in practice the functions necessary to do this for different

Reviewer: Missing a comma
```suggestion
Unfortunately, in practice, the functions necessary to do this for different
```

> plotting libraries can vary substantially as a result of implementation differences, and

Reviewer: Again, two separate (albeit related) ideas in a sentence, so separate them:
```suggestion
plotting libraries can vary substantially as a result of implementation differences. Thus
```

at the time of this writing, IPyOverlay only supports autoupdating connections for
dynamic content inside of ipympl (Matplotlib) and plotly plots. The authors
intend to continue expanding the set of supported visualization libraries for
this particular feature.

<!--These connections are rendered within an `OverlayContainer` as SVG lines
(implementation details)-->

### Context Menus

> A common feature in UI design across most platforms is the context menu.

Reviewer: Might be better to refer to "details on design" as a UI design paradigm/principle in this first sentence where it makes sense rather than wedging it the text later on in the paragraph.

to work well as buttons. Context menus fill this role by hiding these commands until
the user requests them by right-clicking (another form of details on demand) and supporting different menus when
over different components. 
Displaying a context menu similarly requires
rendering some form of widget over another -- IPyOverlay includes a
`ContextMenuArea` component that can wrap any widget within an
`OverlayContainer` and define a context menu with options and event handlers for
that specific component. In the example use case, we add a context menu to
the scatter plot to allow a user to optionally choose between displaying a
violin plot with the distributions versus the density plots (@fig:context_menu).

:::{figure} context_menu.png
:width: 800px
:label: fig:context_menu
An activated context menu providing a selection of commands, which can be
expanded well beyond the single possible action that can be taken from a left-click. 
:::

## Conclusion

The prevalence and utility of Jupyter notebooks in data analysis and scientific
workflows has prompted the development of a community of tools for interactive

> visualization. IPyWidgets, one of the main drivers and the backbone that

Reviewer: I think this sentence is wordy and gets the same idea across without "one of the main drivers and"
```suggestion
visualization. IPyWidgets, the backbone that
```

expanded well beyond the single possible action that can be taken from a left-click. 
:::

### Example Usage

> A minimal code example using IPyOverlay's containers and connection lines is

Reviewer: Thank you for adding the code example!
